lib/publiccloud/instance.pm

```diff
@@ -352,40 +352,65 @@ sub wait_for_ssh {
     my $start_time = time();
     my $check_port = 1;
     my $sleep_period = $args{ignore_wrong_pubkey} ? 20 : 1;
-
+    #
+    my @sem = (1, 1, 1);
+    my @out = ('') x 3;
+    my @command = ('sudo systemctl is-system-running --wait', 'systemd-analyze time', 'sudo journalctl -b | grep -E "Reached target (Cloud-init|Default|Main User Target)"');
+    my @expect = ('running|degraded', 'Startup finished in', 'Reached target ');
+    my $expect_ko_0 = 'maintenance|stopping|offline|unknown';    # wrong states of command[0]
+    my $pass = 0;
     # Looping until reaching timeout or passing two conditions :
     # - SSH port 22 is reachable
-    # - journalctl got message about reaching one of certain targets
+    # - Startup time successfully collected + is-system-running true + journal contains target
+    #   - journalctl got message about reaching one of certain targets
     while ((my $duration = time() - $start_time) < $args{timeout}) {
         if ($check_port) {
             $check_port = 0 if (script_run('nc -vz -w 1 ' . $self->{public_ip} . ' 22', quiet => 1) == 0);
         }
         else {
             # On boottime test we do hard reboot which may change the instance address
             script_run("ssh-keyscan $args{public_ip} | tee -a ~/.ssh/known_hosts") if (get_var('PUBLIC_CLOUD_CHECK_BOOT_TIME'));
-
-            my $output = $self->run_ssh_command(
-                cmd => 'sudo journalctl -b | grep -E "Reached target (Cloud-init|Default|Main User Target)"',
-                proceed_on_failure => 1,
-                username => $args{username});
-            if ($output =~ m/Reached target.*/) {
+            # here 3 check methods 0 to 2 from @command used:
+            foreach my $j (0 .. 2) {
+                $out[$j] = $self->run_ssh_command(cmd => $command[$j], proceed_on_failure => 1, username => $args{username});
```

Comment:
@m-dati checking the output like `$self->run_ssh_command(cmd => $command[$j], proceed_on_failure => 1` is not very good approach. Instead, please try to make use of the awesome work of @grisu48 and his `$self-> ssh_assert_script_run($command[$j])` which will pass if the command exited with `0` as every well-behaved command on Linux does and fail otherwise. Alternatively, there is also `$self-> ssh_script_run()` which will return the exit code for you to process elsewhere.

Reply:
Sorry, this approach was already present in the original version, in master you can still see it. there.
Anyway in this first phase, we still have to choose which UP command to use: the best infrastructure to use could be much more clear after removed the not needed commands.

Reply:
I studied the functions `run_ssh_command` and `ssh_assert_script_run` and I noted that the steps performed are _almost_ identical in both, with the only exception that at the end the first one executes `script_run` while the second one executes `assert_script_run` instead.  
When it comes to `run_ssh_command` and `ssh_script_run` (without assert) the steps performed are _almost_ identical in both the two, with the first one including some parameters providing more flexibilty.

Now, referring to [note-5](https://progress.opensuse.org/issues/126218#note-5), it could occur that the exit code 0 does not fully guarantee target reached, moreover depending on the check type used.
Moreover the ok status could occurr in the loop after some time with result exit>0, that an assertion will soon terminate.  
Finally for check type n.1 I am considering to let it to pass even with an eventual 'degraded' state (exit >0), allowing the test to proceed and check afterward the eventual impact.

Reply:
Maybe instead of loop we can use `ssh_script_retry()`?

Also only `journalctl` requires us to parse the output.
Both `systemctl is-system-running --wait` as well as `systemd-analyze time` should return `0` only if/when the system is booted.